### DISCUSSION TITLE
Future of the versioning system

### DISCUSSION THREAD

**Discussion by Wohlstand** [Wohlstand/TheXTech#353] [Planning]
Hello!

This is a small discussion about the versioning system used by TheXTech. First, I want to explain the current system:
- The first release of TheXTech I started with 1.3 (but 1.3.1 was released publically the first time) to mean that TheXTech is a direct continuation of the original SMBX game which has the 1.3 version being last.
- Version number X.Y.Z.Q:
  - X - constant 1
  - Y - major version 3, also keeps constant
  - Z - minor version (changes every time when significant updates happen or anything other developed with a hard effort)
  - Q - revision/patch level (changes when a minor set of tweaks or insignificant changes happen, often a set of small changes made over several days)

List of versions based on current versioning system:
- 1.3
- 1.3.1
- 1.3.2
- 1.3.2.1
- 1.3.2.2
- 1.3.2.3
- 1.3.2.4
- ...
- 1.3.5.3
- 1.3.6
- etc.

However, as we discussed on Discord, there are several arguments to change the versioning system:
- Technically, TheXTech itself is a true new major version of the original SMBX engine because the entire kernel got been changed (ported from VB6 to C++ with the general change of used technologies), so, technically TheXTech is real "SMBX 2.0" (in term of versioning, please don't confuse with the X2 project which is totally unrelated to this topic). But, the fact it's a brand new project, it should start versioning from 1.0 but starting from 1.3 to mean a direct continuation of the original SMBX engine.
- The "1.3" prefix keeps unchanged mainly to refer to relativity to the original SMBX 1.3 game, BUT, the name of the project (of the engine, not of the game) is no longer "SMBX", so, it shouldn't be confused with other branches of SMBX game (usually SMBX2 that names self as "SMBX 2.0" and SMBX-38A that initially named self as "SMBX 1.4", but, they are totally different engines with nothing common except SMBX64 standard support).
- Instead of increasing the third section of version (Z) for every significant update, there is reason to bump the minor version (Y) instead.

How the versions would go if I used the versioning system instead:
- 1.3
- 1.3.1
- 1.4
- 1.4.1
- 1.4.2
- 1.4.3
- 1.4.4
- ...
- 1.6.3
- 1.7
- etc.

Anyway, switching the versioning systems will mean that the next version after the upcoming 1.3.6 will be 1.7 (basically, I'll shift version sections and bump the minor (Y) version from 3 to 7, and the third section (Z) will be used for minor updates without serious functional changes, and the fourth section (Q) will be used for hotfixes only over already released versions).

So:
- ...
- 1.3.5.3 (stable release without hotfix)
- 1.3.5.3-1 (current stable release with a hotfix applied)
- 1.3.6 (will be the last version that will use the old versioning)
- 1.6.1 (if released as a minor update over 1.3.6, otherwise start from 1.7)
- 1.7

Also, in theory, would be useful to use even/odd version numbers to separate stable and in-development versions. Currently, in in-development versions, I add the "-dev" suffix, which is enough to mark the in-development version states.

What do you think?

**Comment by LoveBodhi** [comment]
SMBX 1.3 is just an asset name. This is NOT a direct SMBX Engine.
For example: My Nostalgic Paradise asset is no longer related to SMBX.

**Comment by 0lhi** [comment]
I think we should stick at 1.3.x as long as there is no support for major 38A and X2 features.

As soon as 38A/X2 episodes can be played natively or ported easily (without tricky workarounds like in Star Scramble/NSMB Classic), it would be proper to have a version number higher than these. Imo, there is no need to rush, since that's more of a longterm thing.

Also, not exactly relevant, but I feel it's still useful to mention here: I think 1.3.7 should be the version to introduce Netplay support. This will skyrocket the popularity of TheXTech instantly, making it likely to attract developers willing to port X2/38A features.

**Comment by LoveBodhi** [comment]
The other thing is, use 1.3.x.y for standard 1.3 features, higher of 1.3 (1.x.y.z) should be used for multi-features.

**Comment by ds-sloth** [comment]
I agree with @0lhi on this -- let's let 1.3.6 be the update with multires, and 1.3.7 be the update with the next major feature (possibly netplay, or settings/compat rework if we see this as major enough). While the base game support remains SMBX 1.3, let's keep the major/minor version 1.3. The main goal of TheXTech is to faithfully reimplement the engine of SMBX, and using the version to indicate the standard we are targeting seems very reasonable to me.

(We could do 1.4, 2.0, or 2.4 in the future if/when we support the appropriate base featuresets.)

**Reply by Wohlstand** [reply]
Also, there are other awaiting or planned add-ons such as a font engine to print non-ASCII (and the translation system for the engine itself, and the non-destructive localization for episodes), lua scripting system, graphical backend update (to allow drawing any texture by Z-order) which is very important for a doze of graphical nifty features and for multi-layer backgrounds (that should be able to draw foreground parts, and draw parts among in-scene objects, not just behind of everything), etc. etc. etc. And so, speaking about standard extensions, they really could bump the major version from 1 to 2, etc., and we should be fine to bump minor version every set of biggies introduced.

Additionally, I think, having a set of biggies in the next time (after porting existing devel-branch stuff), we should choose the plan which will make the next stable release, or which will be taken in the next version. And so, we can release stable versions more often, to give users new features and bugfixes quicker.

**Comment by Wohlstand** [comment]
Speaking about the version, why we need to switch it, as an example of the SDL project, for many years since the release of SDL2, it had versions from 2.0.0 to 2.0.22, with a time it got many internal reworks, major updates, received lots of new functionality, etc. The whole versioning system was confused, the "major" was being a fixed 2.0, when in fact it got lots of new additions of API, etc. But the minor version wasn't changed, and since the recent time they switched the versioning system with the removal of the minor version and so, their next version is 2.23.1, and the next release planned is SDL 2.24. So, they took general versioning rules:
- Major version, changes when any API/ABI incompatibility is introduced (such as removal of deprecated methods, renaming existing functions, replacing API parts with different, etc).
- Minor version changes when API receives new functions and sub-systems, or internal parts got been reworked, updated, etc.
- Revision number changes when internality changes (bug fixing, optimizations, stabilization, etc) without any API updates.
- Patch number, changes when very any small changes were done, like, minor bug fixes, hotfixes, etc.

**Comment by 0lhi** [comment]
How do you feel about doing a 1.3.5.9 Release with Autocode + Abstract Controls, but without Multires?

I suggested this recently in Chat but the Bridge Bot had issues, so I don't know how the idea was received.

**Reply by Wohlstand** [reply]
As we talked at Discord last time, we talked that let 1.3.6 will be a release based on the current state of branch "`master`", and 1.3.7 (or 1.7) will be a future release with multi-res and other nifty stuff.

If we all agree, then I move the rest of the current tasks into the 1.3.7 (or 1.7) milestone and keep only critical bugfix tasks to prepare for the next release.

**Reply by 0lhi** [reply]
Was that agreed on? @ds-sloth and I both think the version number shouldn't exceed 1.3.x until 38A/X2 stuff is supported.

If we do a 1.3.6 release now and give 1.3.7 to Multires, we'll run out of appropriate version numbers in no time.

**Reply by 0lhi** [reply]
> If we all agree, then I move the rest of the current tasks into the 1.3.7 (or 1.7) milestone and keep only critical bugfix tasks to prepare for the next release.

I don't agree with that. I am against 1.3.7 for Multires because then we'll run out of version numbers. And I'm against 1.7 because it can be easily interpreted as "*has surpassed 1.4*", which just isn't the case yet. TheXTech is the best engine to replay old episodes. But it has still has to catch up on offering NPCs + Blocks for new episodes.

1.3.5.9 would be okay if people really want an Autocode/Abstract Controls release now. The jump from .3 to .9 indicates that a big change was made, and that it's almost 1.3.6, but not quite 1.3.6. Imo, that's the best option within current circumstances, other than not making a release until 1.3.6, which I would also be fine with.

**Reply by Wohlstand** [reply]
Why I think about 1.3.6 and 1.3.7 because even now a lot of major changes done in the core and a lot of new significant features made, and 1.3.5.4/5 is too small a version update for such a release.

**Reply by LoveBodhi** [reply]
We SHOULDN'T use non-1.3 versions for a 1.3 Standard only! 1.4 and 2.0 (X2, 2nd generation of SMBX) are NEW FEATURE-based but NOT freestyle!

**Reply by LoveBodhi** [reply]
Versions over 1.3 should feature Freestyle-Asset system!

**Reply by ds-sloth** [reply]
Realistically, the changes in the abstract controls / autocode / editor release are bigger than the ones in most of the previous 1.3.x releases, so I think it would be reasonable to let it be 1.3.6. Let's not focus on running out of numbers, but rather on representing the work that has gone into each release. 1.3.5.9 also works in my view.

After that, we can merge multires and block quadtree into the main branch for testing, with the plan that they will form the core of 1.3.7. (The tree is stable but has lower performance than the normal FLblock array in version conditions. Any performance improvements to the quadtree would not alter gameplay, so we can work on those more slowly.) After plenty of testing, we can have a 1.3.7 release with a very polished version of those features and possibly also the config / compat update, which won't impact gameplay, just make it easier to configure the game, so it won't need as exhaustive of testing. 

I'm becoming convinced that the best way to handle the "running out of versions" issue without indicating that we've surpassed 1.3 might be to actually drop the 1.3 entirely. So, 6.0, 7.0, etc.

**Reply by 0lhi** [reply]
> I'm becoming convinced that the best way to handle the "running out of versions" issue without indicating that we've surpassed 1.3 might be to actually drop the 1.3 entirely. So, 6.0, 7.0, etc.

SMBX2 is known as SMBX 2.0. So, that's not ideal either.

I'm definitely in favor of dropping the 1.3 at some point, but, not yet.

If 1.3.7 will be Multires + Settings rewrite, I guess 1.3.8 would be the target for Netplay then?

Ideally, what would follow after is the stuff discussed in #129 and #239. Those are the point imo where we should get past 1.3.

**Reply by ds-sloth** [reply]
Since 6.0 is far from 2.0, people should understand we're not trying to be directly comparable. Also, it's TheXTech 6.0, not SMBX 6.0, so that should also help?

**Reply by 0lhi** [reply]
Separating TheXTech version number from the asset pack version number is an option. Not my favorite one, but something I would be okay with.

In that case, we should wait until #291 to make the change. Once a ```SMBX.v1.3.6.xta.7z``` is available for download, we can detach the version number of the engine from it.

**Reply by ds-sloth** [reply]
Right, I completely forgot that we have the release that pretends to be an SMBX version. (Before I knew about TheXTech, I downloaded an "SMBX 1.3.2" build that was actually TheXTech.) I think it might be useful to rename the release "SMBX.v1.3.xta.6.0.7z` or something like that?

**Reply by 0lhi** [reply]
Hmm. It's worth considering at least. Maybe ```SMBX.v1.3e.xta.6.0.7z``` to also indicate that it's not the pure Vanilla Assets, but an extended pack. But that should probably be discussed at #291.

**Reply by LoveBodhi** [reply]
For Nostalgic Paradise, version numbering will be like NsPr.0.2.xta.Y.7z.

**Reply by LoveBodhi** [reply]
I see, Nostalgic Paradise 1.0 should feature basic assets (like Completed Tileset, Basic NPC, etc.).

**Comment by LoveBodhi** [comment]
So will be 1.3.10?! (Future Freestyle-version like custom NPC/Block)

**Reply by 0lhi** [reply]
Not a fan of using two digits next to each other :/ I kinda hope Lua support will be added after 1.3.7, which could justify jumping beyond 1.3.x.

But this gives me an idea. Maybe we could change the last digit to a letter?

Current version is 1.3.5.3. I had to look that up cause, who can remember 4 numbers? I can't.

If we make another release before Multires, it could be 1.3.5.d? Would be easier to remember, would offer room for 26 small updates before changing the major version, and it wouldn't look as bad if we do ever get to the point of 1.3.10 (compare ```1.3.10.7``` to ```1.3.10.g```).

**Reply by LoveBodhi** [reply]
Multires is before 1.3.6.
So 1.3.5.x is suitable before that.

**Reply by Wohlstand** [reply]
Anyway, Autocode and Abstract controls are too major features to be a small update of 1.3.5.x branch, and therefore I think, it's reasonable to release 1.3.6 with the current features set and then 1.3.7 with multires and other nifty stuff. Currently multires needs more testing/polishing until get an average stable state to be released in a stable update. We with @0lhi still thinking about how better to do this.

**Comment by Wohlstand** [comment]
Okay, just a small poll, which versioning number should be used after the 1.3.6 release? (react to this message to vote)
- :+1: Use a new versioning model based on **general software versioning** (major, minor, revision, tweak) (the next version will be 1.7):
  - Major - gets changed when introduces any changes that break backwards compatibility (for TheXTech will stay as 1)
  - Minor - gets changed when new significant features get added (will start from 7)
  - Revision - gets changed when a minor set of features, bugfixes or internal optimizations added
  - Tweak - gets used and changed when various hotfixes get released
- :rocket: New alternative versioning (similar to first, but major version will get changed more often (the next version will be 7.0):
  - Major - gets changed when new significant features get added (will start from 7)
  - Minor - gets changed when an average set of features, bugfixes or internal optimizations added
  - Revision - gets changed when making a small bunch of additions without any significant works
  - Tweak - gets used and changed when various hotfixes get released
- :eyes: Stay use the current versioning number (the next version will be 1.3.7):
  - Two version blocks as a major version (1.3)
  - Revision number as a minor version (7)
  - Tweak number as revision version (bugfixes, small features, etc.)
- :heart: - Own suggestion

**Note for new versioning model**: As some people said about possible confusion because of version number shows the engine "surpasses" other engines so quick, it's not to compare. Every branch of SMBX has its own independent versioning model and every SMBX branch is known by a different name, not by version number. So: SMBX 1.3, SMBX-38A or 38A, SMBX2 or X2, and TheXTech. So, the different names of the engine should be enough to understand that it's a different engine with different features set.

**Reply by ds-sloth** [reply]
Can you add an option that uses Major and Minor only? (So, TheXTech 7.0 for example.)

**Reply by Wohlstand** [reply]
Added

**Reply by LoveBodhi** [reply]
My another suggestion is separate with LUA script or not. Remove ".3" after LUA script, retain ".3" before it.

**Reply by 0lhi** [reply]
@LoveBodhi I think there's a strong consensus that we should move away from "1.3.x" eventually.

The question is what we should do for the next versions.

If you want to keep .3 for now (I want the same), you should vote for the 👀 eyes.

**Reply by ds-sloth** [reply]
I’m happy with keeping the 1.3 series for a while longer too. I’m in no rush for this! But it will be nice for us to separately consider what will come next.

**Reply by Wohlstand** [reply]
Anyway, what if I add the basic lua support (with minimal functionality on the quick hand) after the 1.3.6 release? :)

**Reply by LoveBodhi** [reply]
> Anyway, what if I add the basic lua support (with minimal functionality on the quick hand) after the 1.3.6 release? :)

Then .3 will be removed.

**Reply by 0lhi** [reply]
> I’m in no rush for this! But it will be nice for us to separately consider what will come next.

@ds-sloth Agreed. Sort of related, could you join Matrix later? There's an idea I'd like to discuss with you.

> Anyway, what if I add the basic lua support (with minimal functionality on the quick hand) after the 1.3.6 release? :)

@Wohlstand While I'm SUPER eager to get Lua support, let's wait just a bit before making a decision. There's also another other thing to discuss, but best to wait until Merge to prevent things from getting overwhelming ^-^
